src/components/Email/ManageWalletModalWithEmail.tsx

```diff
@@ -0,0 +1,193 @@
+                unsubscribedFromAll
+                unsubscribedFromNotifications
+              }
+            }
+          }
+        }
+
+        ...EmailManagerFragment
+        ...ManageWalletsFragment
+      }
+    `,
+    queryRef
+  );
+
+  const updateEmailNotificationSettings = useUpdateEmailNotificationSettings();
+  const [isEmailNotificationChecked, setIsEmailNotificationChecked] = useState(false);
+  const [isShowAddEmail, setIsShowAddEmail] = useState(false);
+  const { pushToast } = useToastActions();
+  const reportError = useReportError();
+
+  // Invert the value since in the db stored as unsubscribe while in the UI it's subscribe
+  const isEmailNotificationSubscribed =
+    query?.viewer?.email?.emailNotificationSettings?.unsubscribedFromNotifications || false;
+
+  const isEmailUnsubscribedFromAll =
+    !!query?.viewer?.email?.emailNotificationSettings?.unsubscribedFromAll ?? false;
+
+  // Set the initial state of the email notification toggle
+  // so we can instantly toggle it on/off without waiting for the mutation to complete
+  useEffect(() => {
+    setIsEmailNotificationChecked(isEmailNotificationSubscribed);
+  }, [isEmailNotificationSubscribed]);
+
+  // If the user already have email attached, toggle the email manager ui
+  const userEmail = query?.viewer?.email?.email;
+  useEffect(() => {
+    setIsShowAddEmail(Boolean(userEmail));
+  }, [userEmail]);
+
+  const [isPending, setIsPending] = useState(false);
+
+  const handleEmailNotificationChange = useCallback(
+    async (checked: boolean) => {
+      const unsubscribedFromNotifications = !checked;
+      setIsEmailNotificationChecked(checked);
+      setIsPending(true);
+      try {
+        const response = await updateEmailNotificationSettings(
+          unsubscribedFromNotifications,
+          isEmailUnsubscribedFromAll
+        );
+
+        // If its failed, revert the toggle state
+        if (!response?.updateEmailNotificationSettings) {
+          // setIsEmailNotificationChecked(!checked);
+          pushToast({
+            message:
+              'Settings successfully updated. You will no longer receive notification emails',
+          });
+          return;
+        }
+        pushToast({
+          message: 'Settings successfully updated. You will now receive notification emails',
+        });
+        return;
+      } catch (error) {
+        if (error instanceof Error) {
+          reportError('Failed to update email notification settings');
+        }
+
+        setIsEmailNotificationChecked(!checked);
+      } finally {
+        setIsPending(false);
+      }
+    },
+    [isEmailUnsubscribedFromAll, pushToast, reportError, updateEmailNotificationSettings]
+  );
+
+  const toggleEmailNotification = useCallback(() => {
+    // Invert the value from subscribe to unsubscribe
+    handleEmailNotificationChange(!isEmailNotificationChecked);
+  }, [handleEmailNotificationChange, isEmailNotificationChecked]);
```

Comment:
You're passing this callback to `<Toggle onChange>`. That callback seems to pass the new value as a parameter, if we use that parameter instead of relying on `isEmailNotificationChecked`, then `toggleEmailNotification` will be stable and won't cause re-renders.

Reply:
Good idea! Updated to use native checked value instead of `isEmailNotificationChecked`